util/Pos/Util/Verification.hs

```diff
@@ -0,0 +1,101 @@
+{- |
+
+Machinery to support pure datatype verification.
+
+1. Datatypes can have and not have internal pure consistency. If
+datatype doesn't have any implicit rules, it's called checked. 'Integer'
+is checked, 'Either Text String' is checked too. Otherwise, it's called
+unchecked. 'Coin' is unchecked, because it has implicit upper bound on the
+value, which is lower than 'Word64' bound. Newtype over string that is
+supposed to contain only strings of length less than 30 symbols is
+unchecked.
+
+2. Unchecked types' constructors should have prefix \"Unchecked\" (for
+example, 'UncheckedProxySecretKey'). They should also have description
+of why are they unchecked and provide some easy-to-use builders
+(@mkX@) if needed.
+
+3. If unchecked datatype @S@ is a part of @T@ (e.g. record field),
+then @T@ is not called unchecked. In other words, "being unchecked"
+property doesn't propagate, it is only related to the datatype itself.
+
+4. @T@ is called verifiable if it is unchecked or any of its parts are
+verifiable.
+
+5. Every verifiable type must have 'PVerifiable' instance. See
+the description of this class.
+
+-}
+module Pos.Util.Verification
+    (
+      PVer
+    , pverFail
+    , pverField
+
+    , PVerifiable(..)
+    , runPVerify
+    , runPVerifyText
+    , runPVerifyPanic
+    ) where
+
+import           Universum
+
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Text as T
+import           Formatting (sformat)
+import           Serokell.Util (VerificationRes (..), listJson, verResToMonadError)
+
+-- Prefix P is annoying, but we need to distinguish between pure
+-- verification and VAR somehow.
+
+-- | Pure verification monad. It accumulates errors along the
+-- verification (AST) path.
+newtype PVer a = PVer
+    { getPVer :: Either (NonEmpty Text) a
+    } deriving (Show, Eq, Functor, Applicative, Monad)
+
+-- | Fail inside the 'PVer' monad.
+pverFail :: Text -> PVer ()
+pverFail t = PVer (Left $ one t)
+
+-- | Verifies some field, prefixing with the text value in case of
+-- error. Prefix is supposed to be the record field name. Use it when
+-- you want to specify which field/component are you verifying.
+pverField :: Text -> PVer () -> PVer ()
+pverField p (PVer v) = PVer $ first (\x -> p `NE.cons` x) v
+
+-- | Things that can be (purely) verified.
+class PVerifiable a where
+
+    -- | Verify the 'a' predicates only, assuming that all subfields
+    -- are correct. Is needed to be implemented for unsafe types only.
+    pverifySelf :: a -> PVer ()
+
+    -- | Verify all subcomponents and the component itself. Basically,
+    -- call 'pverify' for subcomponents and 'pverifySelf' in the end.
+    pverify :: a -> PVer ()
```

Comment:
Wouldn't it make more sense to have `pverifySelf` and `pverifyParts` and then `pverify a = pverifyParts a >>= pverifySelf a` (outside this class probably)? This way one wouldn't have to call `pverifySelf` in definition of `pverify`.

Reply:
Ok, fixed.

Reply:
It's not that trivial as you say, but I thought again and I think I have an idea.

Reply:
I introduced a `PVerifiableSub` wrapper which solves the problem you are describing.